# Exercice une mouche et deux trains

Deux trains, séparés de 200 km roulent l'un vers l'autre. Chacun avance à 50 km/h. 
Une mouche part de l'avant de l'un d'eux et vole à la vitesse de 75 km/h jusqu'à ce qu'elle rencontre le second train. A ce moment, elle fait demi-tour, jusqu'à ce qu'elle rencontre le premier train, puis fait demi-tour jusqu'à ce qu'elle rencontre le second et ainsi de suite, jusqu'à ce que les trains la tuent en se croisant. 

Faire un notebook python pour simuler les va et viens de la mouche et trouver le resultat.  
Bonus: expliquer pourquoi il ne sers à rien.

In [1]:
# Conditions initiales
time = 0
T1 = 0
T2 = 200
M = 0
VT1 = VT2 = 50
VM = 75
distance = 0

In [2]:
# fonction qui calcule la durée du trajet de la mouche sur une itération
def duree_mouche():
    global T1, T2
    d = (T2 - T1) / (VM+VT1)
    return d

In [3]:
# fonction qui met à jour les positions des trains et de la mouche après un trajet d'une durée d
def new_positions(d):
    global T1, T2, distance
    T1 = T1 + (d*VT1)
    T2 = T2 - (d*VT2)
    distance = distance + (VM*d)

In [4]:
# Programme principal

iter = int(input("Combien d\'itérations voulez-vous?: ")) 
for i in range(iter):
    print('itération : ', i+1)
    new_positions(duree_mouche())
    print('Position T1:',T1,' Position T2:', T2, 'Distance parcourue par la mouche:', distance)


Combien d'itérations voulez-vous?: 5
itération :  1
Position T1: 80.0  Position T2: 120.0 Distance parcourue par la mouche: 120.0
itération :  2
Position T1: 96.0  Position T2: 104.0 Distance parcourue par la mouche: 144.0
itération :  3
Position T1: 99.2  Position T2: 100.8 Distance parcourue par la mouche: 148.8
itération :  4
Position T1: 99.84  Position T2: 100.16 Distance parcourue par la mouche: 149.76000000000002
itération :  5
Position T1: 99.968  Position T2: 100.032 Distance parcourue par la mouche: 149.95200000000003


# Rappel sur les suites géométriques
https://www.maths-et-tiques.fr/telech/SuitesAG.pdf

## Définition

u _(n+1)_ = q × u _(n)_

## Propriétés

u _(n)_ = u _(0)_ × q^n

n est un entier naturel non nul et q un réel différent de 1 alors on a :
1+q+q^2+...+q^n = (1−q^(n+1))/(1-q))

# La distance totale parcourue par la mouche est une suite géométrique
En effet, on observe qu'à chaque itération n, la mouche parcourt 0,2 fois la distance qu'elle a parcourue à l'itération n-1.  On peut donc en déduire :

## Suite géométrique distance mouche
d0 = 120  
d1 = 24 = 0,2 * d0  
d2 = 4.8 = 0,2 * d1  
...  
d(n+1) = 0,2 * dn

## Distance totale parcourue 
d_totale = d0 * (1-0,2^infini) / (1-0,2)  
= 120 * 1 / 0,8  
= 150 Km